ADP-455: return unsigned delegation certificate

hasufell opened the pull request:
# Issue Number

#2200 

# Overview

- [x] Added Api types
- [x] Refactored some of the Transaction layer (still WIP)
- [x] Factored out certificate creation from `joinStakePool`
- [x] Added handler for joining stake pool
- [x] Add handler for quitting stake pool
- [x] serialization of certificates
- [x] lots of cleanup
- [x] fix jormungandr
- [x] fix tests
- [ ] add tests


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1954,10 +1956,9 @@ joinStakePool
     -> Passphrase "raw"
-    -> ExceptT ErrJoinStakePool IO (Tx, TxMeta, UTCTime)
-joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
+    -> ExceptT ErrJoinStakePool IO (CoinSelection, [Cardano.Certificate], DelegationAction)
```
In principle, we don't want to expose a type from `cardano-ledger-specs` or `cardano-api` in here. The code in Cardano.Wallet is supposed to be "backend-agnostic" and this `Cardano.Certificate` is specific to cardano-node (makes no sense for Jormungandr). 

Plus, isn't that redundant with the `DelegationAction` already ?

hasufell replied:
How so? We want to serialise the certificates. So we have to know what types they are: https://input-output-hk.github.io/ouroboros-network/shelley-spec-ledger/Shelley-Spec-Ledger-Delegation-Certificates.html#t:DelegCert

hasufell replied:
In addition, this endpoint won't be available for jormungandr anyway. How would we abstract over this?

KtorZ replied:
But the `DelegationAction` are the certificates already. They capture the exact same concept.

hasufell replied:
So do you mean we keep:

```hs
    _mkDelegationCertificates
        :: DelegationAction
            -- Pool Id to which we're planning to delegate
        -> XPub
            -- Reward account public key
        -> [Cardano.Certificate]
    _mkDelegationCertificates da accXPub =
        case da of
           Join poolId ->
                   [ toStakePoolDlgCert accXPub poolId ]
           RegisterKeyAndJoin poolId ->
                   [ toStakeKeyRegCert  accXPub
                   , toStakePoolDlgCert accXPub poolId
                   ]
           Quit -> [toStakeKeyDeregCert accXPub]
```

private and not in TransactionLayer and then use it when we serialize DelegationAction in the shelley API?

KtorZ replied:
Yes. This is a shelley-specific detail so it shouldn't escape the shelley package.

hasufell replied:
:godmode:

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1976,21 +1977,62 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
+            let pwdP = preparePassphrase scheme pwd
+            let rewardAcnt = getRawKey $ deriveRewardAccount @k pwdP xprv
+            let certs = (mkDelegationCertificates tl) (W.Join pid)
+                    (rewardAcnt, pwdP) cs
```
We should _not_ require the passphrase nor the private key for this. Actually, the main use-case for such function is when used for hardware devices integration in which case we do not have access to the root private key whatsoever. 

It is okay to enforce that `s ~ SeqState n k` here because it makes no sense anyway to call this for a `RndState` (random wallets) which have no delegation capabilities. From there, it is possible to access the `rewardAccountKey` corresponding to that particular account without any need to access the root private key.

hasufell replied:
We also need `k ~ ShelleyKey` afais, for `getKey`.

KtorZ replied:
You can use `WalletKey k` as a constraint and use `getRawKey`

hasufell replied:
```
src/Cardano/Wallet/Jormungandr/Api/Server.hs:191:13: error:
    • Couldn't match type ‘JormungandrKey’
                     with ‘Cardano.Wallet.Primitive.AddressDerivation.Shelley.ShelleyKey’
        arising from a use of ‘joinStakePool’
    • In the first argument of ‘(:<|>)’, namely
        ‘joinStakePool
           jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)’
      In the second argument of ‘(:<|>)’, namely
        ‘joinStakePool
           jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)
           :<|> quitStakePool jormungandr :<|> delegationFee jormungandr’
      In the expression:
        (listPools spl)
          :<|>
            joinStakePool
              jormungandr (knownStakePools spl) (getPoolLifeCycleStatus spl)
              :<|> quitStakePool jormungandr :<|> delegationFee jormungandr
    |
191 |             joinStakePool jormungandr
```

hasufell replied:
nvm, found it

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -68,9 +72,21 @@ data TransactionLayer t k = TransactionLayer
+        :: W.DelegationAction
+            -- Pool Id to which we're planning to delegate
+        -> (XPrv, Passphrase "encryption")
+            -- Reward account
```
This should be passed as a public key, not an encrypted private key.

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -68,9 +72,21 @@ data TransactionLayer t k = TransactionLayer
+            -- Pool Id to which we're planning to delegate
+        -> (XPrv, Passphrase "encryption")
+            -- Reward account
+        -> CoinSelection
```
Our current logic is a bit convoluted / backward mostly because of the Jormungandr support. In principle, we construct the coin selection from the certificates, not the other way around. We can indeed _infer_ what certificates were assumed for the coin-selection but that information is already solely available from `DelegationAction`. 

RegisterKeyAndJoin -> a stake key certificate + stake delegation certificate
Join -> a stake delegation certificate
Quit -> a stake key de-registration certificate

hasufell replied:
Yeah, I saw that and I was feeling a bit odd about it. Also because we have two `DelegationAction` types now... one just `Join | Quit` (for the API input) and the other what you wrote above.

KtorZ replied:
It'd be okay to use `ApiT DelegationAction` in the API, and have the JSON instance fails on parsing `RegisterKeyAndJoin` so that only `Join` and `Quit` are exposed in the JSON interface.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -997,11 +1016,42 @@ instance DecodeAddress n => FromJSON (ApiSelectCoinsData n) where
+    toJSON (ApiSelectCoinsActionData Quit pass)
+        = object [ "action" .= String "quit", "passphrase" .= pass ]
+    toJSON (ApiSelectCoinsActionData (Join poolid) pass)
+        = object [ "action" .= String "join", "pool" .= ApiT poolid, "passphrase" .= pass ]
```
These would be better derived using Generics. Hand-written JSON usually leads to inconsistencies and are hard to maintain on large APIs.

hasufell replied:
The representation is too different from the data IMO and I believe that's what json roundtrip tests are for.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1712,9 +1764,15 @@ type instance ApiAddressIdT (n :: NetworkDiscriminant) =
     ApiSelectCoinsData n
 
+type instance ApiSelectCoinsActionDataT (n :: NetworkDiscriminant) =
+    ApiSelectCoinsActionData n
```
I don't think we need such type instances here. We need those for data-types which carry _addresses_, because of the type-level network discriminant. But if the underlying data-type do not contain any addresses, then this is superfluous / more complex than necessary.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -402,6 +408,19 @@ newtype ApiSelectCoinsData (n :: NetworkDiscriminant) = ApiSelectCoinsData
+data ApiCoinSelectionCert (n :: NetworkDiscriminant) = ApiCoinSelectionCert
+    { inputs :: !(NonEmpty (ApiCoinSelectionInput n))
+    , outputs :: !(NonEmpty (AddressAmount (ApiT Address, Proxy n)))
+    , certs :: ! [ApiT Certificate]
```
~~certs~~ -> `certificates`, bytes are cheap in the API :)

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -402,6 +408,19 @@ newtype ApiSelectCoinsData (n :: NetworkDiscriminant) = ApiSelectCoinsData
+    { inputs :: !(NonEmpty (ApiCoinSelectionInput n))
+    , outputs :: !(NonEmpty (AddressAmount (ApiT Address, Proxy n)))
+    , certs :: ! [ApiT Certificate]
+    } deriving (Eq, Generic, Show)
```
What about simply extending the current `ApiCoinSelection` data-type? The list of certificates can be empty for other delegation and we need not to define yet-another-resource.

hasufell replied:
Isn't the point of types to have clear separation? For the old endpoint, `certs` will always be null. That should be expressed in the type system, so a potential client knows it doesn't have to deal with it.

KtorZ replied:
> For the old endpoint,

That's the point though, it is the _same_ endpoint, manipulating the same resource. Just a different request payload. It would indeed be always empty for normal payments, and that is just right because normal payments don't have any certificates. The introduction of certificates done by this PR is an extension of the existing resource, not a new definition.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -402,6 +408,19 @@ newtype ApiSelectCoinsData (n :: NetworkDiscriminant) = ApiSelectCoinsData
+data ApiSelectCoinsActionData (n :: NetworkDiscriminant)
+    = ApiSelectCoinsActionData
+    { action :: W.DelegationAction
+    , passphrase :: ApiT (Passphrase "raw")
```
As outlined in other comments, the `passphrase` should not be required.

hasufell replied:
I am totally out of my depth on this one.

KtorZ commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -304,6 +307,13 @@ type SelectCoins n = "wallets"
+    :> "coin-selections"
+    :> "random"
+    :> ReqBody '[JSON] (ApiSelectCoinsActionDataT n)
+    :> Post '[JSON] (ApiCoinSelectionCertT n)
```
Interesting. Does `servant` treats `:<|>` as a type-level Alternative now? I am pretty sure in the past that only the right-hand side would be picked by servant when used like this (with identical "signatures" on endpoints). If that acts as a sort-of type-level alternative then that's quite amazing and there are a few other API endpoints we can simplify :thinking:

hasufell replied:
https://hackage.haskell.org/package/servant-0.18/docs/Servant-API.html#t::-60--124--62-

Commit:
Unsigned delegation

hasufell commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -313,6 +314,13 @@ type SelectCoins n = "wallets"
     :> ReqBody '[JSON] (ApiSelectCoinsDataT n)
     :> Post '[JSON] (ApiCoinSelectionT n)
 
+type SelectCoinsAction n = "wallets"
```
I forgot if we want to keep this or extend `ApiSelectCoinsData` instead?

```hs
newtype ApiSelectCoinsData (n :: NetworkDiscriminant) = ApiSelectCoinsData
    { payments :: NonEmpty (AddressAmount (ApiT Address, Proxy n))
    , action :: Maybe DelegationAction
    } deriving (Eq, Generic, Show)
```

something like that?

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1688,47 +1692,18 @@ selectCoinsExternal ctx wid argGenChange payments withdrawal md = do
-    | ErrSelectCoinsExternalUnableToAssignInputs WalletId
-    | ErrSelectCoinsExternalUnableToAssignOutputs WalletId
+    | ErrSelectCoinsExternalUnableToAssignInputs ErrNoSuchWallet
+    | ErrSelectCoinsExternalUnableToAssignOutputs ErrNoSuchWallet
```
I don't understand the rationale here? The `ErrNoSuchWallet` refers to an action which failed because there was no wallet associated with a given id. The `UnableToAssignInputs` and `UnableToAssignOutputs` errors supposes the existence of a wallet because they happened after a coin selection was found on a given wallet! 

Arguably, the `WalletId` in these errors brings very little information in the first place. It'd be more interesting to carry the actual malformed coin-selection in the error context to help debugging.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2494,6 +2601,39 @@ guardCoinSelection minUtxoValue cs@CoinSelection{outputs, change} = do
+    -> ExceptT
+        e
+        IO
+        (NonEmpty (TxIn, TxOut, NonEmpty DerivationIndex))
```
I don't think we need so many newlines here if everything can fit on one line :+1:

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2494,6 +2601,39 @@ guardCoinSelection minUtxoValue cs@CoinSelection{outputs, change} = do
+        -> ExceptT
+            e
+            IO
+            (TxIn, TxOut, NonEmpty DerivationIndex)
```
Same here.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1972,32 +1947,58 @@ getTransaction ctx wid tid = db & \DBLayer{..} -> do
-    -> Passphrase "raw"
-    -> ExceptT ErrJoinStakePool IO (Tx, TxMeta, UTCTime)
-joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
+    -> ExceptT ErrJoinStakePool IO (UnsignedTx (TxIn, TxOut, NonEmpty DerivationIndex), DelegationAction, [DerivationIndex])
```
We could possibly use an intermediate type `newtype RewardAccountPath = RewardAccountPath [DerivationIndex]` just for the sake of readability.

hasufell replied:
Honestly, I'm not too much of a fan of newtypes for readability. We can just use a type synonym or document it. newtypes don't give much type safety, unless they use smart constructors.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1972,32 +1947,58 @@ getTransaction ctx wid tid = db & \DBLayer{..} -> do
+    -> PoolLifeCycleStatus
+    -> WalletId
+    -> ExceptT ErrJoinStakePool IO (CoinSelection, DelegationAction, [DerivationIndex])
+joinStakePoolUnsigned' ctx currentEpoch knownPools pid poolStatus wid =
```
Why is this split over two functions? The names of the function are not capturing quite well what they are for and the split seems rather arbitrary?

hasufell replied:
Because they return different types, but basically do the same thing. I can't go back to CoinSelection from `UnsignedTx`.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2014,9 +2015,52 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
+            $ readCheckpoint (PrimaryKey wid)
+        let s = getState cp
+            dprefix = Seq.derivationPrefix s
+            sPath = stakePath dprefix
```
We should avoid re-capturing a checkpoint that was already captured earlier in the function handler. By doing this, we may create inconsistency (because we can only ensure that the database is not changed by a concurrent thread within `atomically`). The wallet is already fetched earlier `readWalletMeta` and within the `joinStakePoolUnsigned` function.

hasufell replied:
`readWalletMeta` doesn't return the checkpoint or the state. Can you be a bit more specific?

KtorZ replied:
Instead of fetching _just_ the metadata with `readWalletMeta`, you could fetch the wallet state + metadata, but more importantly, do the coin selection and post-processing within the same `atomically`. Otherwise, there might be another thread interleaving changing the state of the wallet after you ran the coin selection. This is pretty minor here I recall because you're not writing anything back, but that's a potential source of subtle bug. 

Said differently, we should avoid:

```
atomically readCheckpoint
// stuff
atomically writeCheckpoint / readCheckpoint again
// stuff
```

but instead do

```
atomically $ do
   readCheckpoint
   // stuff
   writeCheckpoint 
   // stuff
```

hasufell replied:
Does this fix it?
```hs
joinStakePoolUnsigned' ctx currentEpoch knownPools pid poolStatus wid =
    db & \DBLayer{..} -> do
        (wal, walMeta, _) <- withExceptT
            ErrJoinStakePoolNoSuchWallet (readWallet @ctx @s @k ctx wid)
        isKeyReg <- mapExceptT atomically
            $ withExceptT ErrJoinStakePoolNoSuchWallet
            $ isStakeKeyRegistered (PrimaryKey wid)

        let mRetirementEpoch = view #retirementEpoch <$>
                W.getPoolRetirementCertificate poolStatus
        let retirementInfo =
                PoolRetirementEpochInfo currentEpoch <$> mRetirementEpoch

        withExceptT ErrJoinStakePoolCannotJoin $ except $
            guardJoin knownPools (walMeta ^. #delegation) pid retirementInfo

        let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
        liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg

        cs <- withExceptT ErrJoinStakePoolSelectCoin $
            selectCoinsForDelegation @ctx @s @t @k ctx wid action

        let s = getState wal
            dprefix = Seq.derivationPrefix s
            sPath = stakeDerivationPath dprefix

        pure (cs, action, sPath)
```

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2331,6 +2432,8 @@ data ErrStartTimeLaterThanEndTime = ErrStartTimeLaterThanEndTime
     = ErrSelectForDelegationNoSuchWallet ErrNoSuchWallet
     | ErrSelectForDelegationFee ErrAdjustForFee
+    | ErrSelectForDelegationUnableToAssignInputs ErrNoSuchWallet
+    | ErrSelectForDelegationUnableToAssignOutputs ErrNoSuchWallet
```
Same as above, the `ErrNoSuchWallet` is definitely wrong in this case.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1778,13 +1824,27 @@ rndStateChange ctx (ApiT wid) pwd =
+        (fmap (uncurry mkCertificates) mcerts)
   where
+    mkCertificates :: DelegationAction -> [DerivationIndex] -> [Api.ApiCertificate]
+    mkCertificates action (s:sx) =
```
What about using a `NonEmpty DerivationIndex` from the start and carrying it around?

hasufell replied:
That makes sense, since `stakePath` is a static non-empty list

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1778,13 +1824,27 @@ rndStateChange ctx (ApiT wid) pwd =
+                , Api.JoinPool apiStakePath (ApiT pid)
+                ]
+            Quit-> [Api.QuitPool apiStakePath]
+    mkCertificates _ _ = []
```
This is silently swallowing an error. We should never ever have an empty derivation path here. Having one is clearly a sign that something went really wrong in the internal logic and should be loudly visible / should crash the application.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2137,18 +2197,8 @@ instance Buildable e => LiftHandler (ErrSelectCoinsExternal e) where
-                , toText wid
-                ]
+        ErrSelectCoinsExternalUnableToAssignInputs e -> handler e
+        ErrSelectCoinsExternalUnableToAssignOutputs e -> handler e
```
The previous error was actually correct.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -372,18 +378,49 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
+--
+newtype ApiRelativeDerivationIndex = ApiRelativeDerivationIndex
+    { unApiRelativeDerivationIndex :: Word31
+    } deriving (Bounded, Enum, Eq, Generic, Show)
```
It seems to me that this should not be there? Wrong merge resolution?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -372,18 +378,49 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
-newtype ApiSelectCoinsData (n :: NetworkDiscriminant) = ApiSelectCoinsData
+data ApiSelectCoinsData (n :: NetworkDiscriminant)
+    = ApiSelectForPayment (ApiSelectCoinsPayments n)
+    | ApiSelectForAction ApiSelectCoinsAction
```
I think: 

ForAction -> ForDelegation 

would better capture the idea behind?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -372,18 +378,49 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
     } deriving (Eq, Generic, Show)
 
+newtype ApiSelectCoinsAction = ApiSelectCoinsAction
+    { delegation_action :: ApiT DelegationAction
```
`delegation_action` -> `delegationAction`, this should use camelCase.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -372,18 +378,49 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
+        }
+    | QuitPool
+        { reward_account_path :: NonEmpty (ApiT DerivationIndex)
+        }
```
Fields should also be camel-cased here :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -372,18 +378,49 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
 data ApiCoinSelection (n :: NetworkDiscriminant) = ApiCoinSelection
     { inputs :: !(NonEmpty (ApiCoinSelectionInput n))
     , outputs :: !(NonEmpty (AddressAmount (ApiT Address, Proxy n)))
+    , certificates :: Maybe [ApiCertificate]
```
Wouldn't `Maybe (NonEmpty ApiCertificate)` be a slightly better type for this? There cannot be in principle a `Just []`. Either there are no certificates, or there are one or more.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -991,16 +1029,69 @@ instance FromJSON ApiEpochInfo where
+        p <- o .:? "payments"
+        a <- o .:? "delegation_action"
+        case (p, a) of
+            (Just _, Just _) -> fail "Specified both payments and action, pick one"
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -991,16 +1029,69 @@ instance FromJSON ApiEpochInfo where
+      , sumEncoding = TaggedObject
+          {
+            tagFieldName = "delegation_type"
+          , contentsFieldName = "contents"
```
The `delegation_` prefix seems wrong to me here. This is describing the type of "certificate", not delegation. There's no multiple types of delegation actually. What about:

```hs
{ tagFieldName = "type" -- or "certificate_type"
, contentsFieldName = "details"
}
```

 ?

hasufell replied:
`contentsFieldName` isn't used

KtorZ replied:
Ah! What about indicating that in a comment then :blush: ?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -991,16 +1029,69 @@ instance FromJSON ApiEpochInfo where
+                pid <- o .: "pool"
+                pure (ApiT $ Join (getApiT pid))
+            "quit" -> pure $ ApiT Quit
+            val -> fail ("Unexpeced action value: " <> T.unpack val)
```
Do not spit back invalid values to users, instead, show what are expected values :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -991,16 +1029,69 @@ instance FromJSON ApiEpochInfo where
+            val -> fail ("Unexpeced action value: " <> T.unpack val)
+
+instance ToJSON (ApiT DelegationAction) where
+    toJSON (ApiT (RegisterKeyAndJoin _)) = error "RegisterKeyAndJoin not valid"
```
That we won't have `RegisterKeyAndJoin` as a possible input in the API should not leak up here. May we not make this function partial and simply have a valid instance here :pray: 

The application logic should handle this properly. (or we should use a different type, more constrained that doesn't make it possible).

hasufell replied:
> That we won't have RegisterKeyAndJoin as a possible input in the API should not leak up here.

Why not? This instance is for `ApiT DelegationAction`, not `DelegationAction`. So it's about our API.

KtorZ replied:
We want to avoid partial functions when we can. There's no reason for this function to be partial. That the application logic (even only in the API) do not accept `RegisterKeyAndJoin` as a constructor should not influence the JSON instance itself. We are actually _abusing_ this type in the first place here in the application logic. The proper way to address this would be to have an actual type that captures only the allowed actions; because it's convenient we use the internal type `DelegationAction` wrapped in an `ApiT` but let's not shoot ourselves in the foot on top by making the JSON instances partial

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -203,6 +211,21 @@ utxoInternal = toEnum $ fromEnum UTxOInternal
+
+-- | Full path to the stake key. There's only one.
+stakePath :: DerivationPrefix -> [DerivationIndex]
+stakePath (DerivationPrefix (purpose, coin, acc)) =
```
neat-picking here, but what about: `stakeDerivationPath`

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -203,6 +211,21 @@ utxoInternal = toEnum $ fromEnum UTxOInternal
 mutableAccount = toEnum $ fromEnum MutableAccount
 
+zeroAccount :: Index 'Soft 'AddressK
+zeroAccount = toEnum 0
```
It'd be better to rely on the `Bounded` instance here to avoid hard-coding magic numbers :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -46,6 +47,8 @@ import Data.Text
 import Data.Word
     ( Word16, Word8 )
 
+import GHC.Generics
```
Avoid implicit imports, this ages badly and makes it hard to figure out where variable comes from for future reader. Prefer explicit imports :+1:

hasufell replied:
You can use haskell-language-server to figure out where variables come from. This rule is now obsolete. But I'll comply regardless.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs:
```diff
@@ -847,12 +847,31 @@ instance Malformed (BodyParam (ApiSelectCoinsData ('Testnet pm))) where
+         jsonValid = (first (BodyParam . Aeson.encode) <$> paymentCases) <> jsonValidAction
+         jsonValidAction = first (BodyParam . Aeson.encode) <$>
+            [ ( [aesonQQ| { "action": "join" }|]
+              , "Error in $: No valid parse"
```
The error is not quite satisfactory here, can we come up with something more meaningful which would help user getting to a resolution?

Commit:
Fix review suggestions

Commit:
Assign change addresses in `joinStakePoolUnsigned`

Commit:
Avoid capturing checkpoint multiple times

Commit:
Assign change addresses in `quitStakePoolUnsigned`

Commit:
Add integration tests

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -2892,7 +2947,7 @@ paths:
         content:
           application/json:
-            schema: *ApiSelectCoinsData
+            schema: *ApiByronSelectCoinsData
```
:+1:

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -530,6 +530,41 @@ x-transactionOutputs: &transactionOutputs
+
+x-rewardAccountPath: &rewardAccountPath
+  type: array
+  minItems: 1
```
we could actually even go further and say:

```
minItems: 5
maxItems: 5
```

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -530,6 +530,41 @@ x-transactionOutputs: &transactionOutputs
+  type: array
+  minItems: 1
+  items:
+    type: string
```
Can't this re-use the `derivationIndex` schema instead?

hasufell replied:
There is no such schema?

KtorZ replied:
```
x-derivationSegment: &derivationSegment
  description: An individual segment within a derivation path.
  type: string
  example: 1852H

x-derivationPath: &derivationPath
  description: A path for deriving a child key from a parent key.
  type: array
  minItems: 1
  items: *derivationSegment
```

*derivationSegment / derivationPath.

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -530,6 +530,41 @@ x-transactionOutputs: &transactionOutputs
       amount: *transactionAmount
 
+x-delegationAction: &delegationAction
+  description: A delegation action
```
Perhaps add a comment saying that `pool` is only required for `join` ?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -265,19 +282,17 @@ newTransactionLayer networkId = TransactionLayer
         let payload = TxPayload Nothing certs mkWits
-        mkTx networkId payload tip acc keyFrom cs
+        let ttl = defaultTTL tip
+        mkTx networkId payload ttl acc keyFrom cs
```
Is this the result of a wrong merge-conflict resolution? I don't why this PR would have to change anything regarding transaction TTL?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -172,25 +174,40 @@ instance TxWitnessTagFor IcarusKey where
     -> TxPayload Cardano.Shelley
     -> SlotNo
-    -- ^ Tip of chain, for calculating TTL
+    -- ^ Time to Live
```
Same here, this change looks _fishy_. I believe this is a wrong merge-conflict resolution?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -172,25 +174,40 @@ instance TxWitnessTagFor IcarusKey where
+            (toChimericAccountRaw . toXPub $ rewardAcnt)
+            (withdrawal cs)
+        unsigned = mkUnsignedTx timeToLive cs md wdrls certs
+    in (unsigned, wdrls)
```
What does this function buys us exactly :thinking: ?

hasufell replied:
It's quite neat :smile: 

I believe the idea was originally to re-use it somewhere, but we don't...

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -109,6 +109,8 @@ import qualified Cardano.Api.Typed as Cardano
 import qualified Cardano.Crypto as CC
 import qualified Cardano.Crypto.Hash.Class as Crypto
+import Cardano.Crypto.Wallet
+    ( XPub )
```
neat-picking: this import should not be there. We group imports as:

Prelude
Explicit imports
Qualified imports

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -285,10 +307,12 @@ server byron icarus shelley spl ntp =
+        withLegacyLayer wid
+            (byron, liftHandler $ throwE ErrNotASequentialWallet)
+            (icarus, selectCoins icarus (const $ paymentAddress @n) wid x)
+    byronCoinSelections _ _ = throwError err400
```
What about having a nicer error message here? Something that tells clients that there's no such thing as selecting coins for delegation on a byron wallet because such wallets have no delegation capabilities whatsoever.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1559,7 +1581,22 @@ instance ToSchema (ApiPutAddressesData t) where
+            & properties .~ mconcat
+                [ paymentData ^. properties
+                , actionData ^. properties
+                ]
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs:
```diff
@@ -847,12 +847,31 @@ instance Malformed (BodyParam (ApiSelectCoinsData ('Testnet pm))) where
+         jsonValid = (first (BodyParam . Aeson.encode) <$> paymentCases) <> jsonValidAction
+         jsonValidAction = first (BodyParam . Aeson.encode) <$>
+            [ ( [aesonQQ| { "action": "join" }|]
+              , "Error in $: No valid parse for ApiSelectCoinsAction or ApiSelectCoinsAction"
```
An interesting error. Sounds like a copy-paste that went wrong?

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -203,6 +213,21 @@ utxoInternal = toEnum $ fromEnum UTxOInternal
+    , fromIndex zeroAccount]
+  where
+    fromIndex :: Index t l -> DerivationIndex
+    fromIndex (Index ix) = DerivationIndex ix
```
what about simply using `getIndex` that comes with the `Index` data-type ?

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2483,6 +2550,14 @@ instance LiftHandler ErrQuitStakePool where
+                , pretty e]
+        ErrQuitStakePoolUnableToAssignOutputs e ->
+            apiError err403 UnableToAssignInputOutput $ mconcat
+                [ "Unable to assign outputs from coin selection: "
```
How is that a client error? This is a server error and it should not happen under any circumstances.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2137,18 +2198,14 @@ instance Buildable e => LiftHandler (ErrSelectCoinsExternal e) where
+        ErrSelectCoinsExternalUnableToAssignOutputs e ->
+            apiError err403 UnableToAssignInputOutput $ mconcat
+                [ "Unable to assign outputs from coin selection: "
+                , pretty e]
```
Two remarks here:

a) This is changing the behavior from a server error to a client error. We aren't necessarily against API breaking changes in principle but they should be backed with a rationale. I don't think this is right here, there's nothing that clients can do about this, getting in this situation is a programmer error meaning that we screwed up.

b) May we please keep the first-person tone to remain consistent with the rest of the server errors. It is much more friendly to get errors reported in the first-person :)

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1778,13 +1824,28 @@ rndStateChange ctx (ApiT wid) pwd =
+            RegisterKeyAndJoin pid ->
+                Api.RegisterRewardAccount apiStakePath :|
+                    [Api.JoinPool apiStakePath (ApiT pid)]
+            Quit-> Api.QuitPool apiStakePath :| []
```
readability suggestion:

```hs
mkCertificates
    :: NonEmpty DerivationIndex
    -> DelegationAction
    -> NonEmpty Api.ApiCertificate
mkCertificates xs =
    \case
        Join pid -> NE.fromList
            [ Api.JoinPool apiStakePath (ApiT pid) 
            ]
  
        RegisterKeyAndJoin pid -> NE.fromList
            [ Api.RegisterRewardAccount apiStakePath 
            , Api.JoinPool apiStakePath (ApiT pid)
            ]
  
        Quit -> NE.fromList
            [ Api.QuitPool apiStakePath 
            ]
  where
    apiStakePath = ApiT <$> xs
```

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -552,6 +555,137 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
+                    , expectField #inputs (`shouldSatisfy` (not . null))
+                    , expectField #outputs (`shouldSatisfy` (not . null))
+                    , expectField #certificates (`shouldSatisfy` (not . null))
+                    ]
```
We could go further in the assertion and make sure that there are 2 certificates, one about joining, and one about the key registration.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -552,6 +555,137 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
+                    [ expectResponseCode HTTP.status200
+                    , expectField #inputs (`shouldSatisfy` (not . null))
+                    , expectField #outputs (`shouldSatisfy` (not . null))
+                    , expectField #certificates (`shouldSatisfy` (not . null))
```
Same here, we could assert that the certificate is indeed of type "quit_pool" (and that there's no other certificates present)

Commit:
Apply further review suggestions

Commit:
Fix 'Arbitrary ApiCertificate'

KtorZ committed:
remove duplicated logic with regards to selections for join/quit

  The code was becoming quite convoluted here with function names that
  were screaming for refactoring (when we start adding `'` to functions,
  it's usually a good sign that something can be simplified).

  The main change is actually in 'selectCoinsExternal' which is now
  parameterized over the selection to run. This way, the steps of
  assigning change addresses are factored out in this function and a lot
  of the duplication goes away.

  Another important change is that I've moved the signing and submission
  of join/quit outside of the body of the function. So that each step
  can be ran independently. This avoid the need for weird intermediate product
  types aggregating more and more information. Now, both functions are
  returning a 'DelegationAction' and merely checking that a join or quit
  is possible.

KtorZ committed:
return structured error for invalid wallets, instead of plain string

KtorZ committed:
Merge pull request #2254 from input-output-hk/KtorZ/2200/simplifications

select coins for delegation simplifications
